### PR TITLE
KDP-1761: Upgrade to Pydantic v2

### PR DESCRIPTION
**Pull request by PaulVanSchayck**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,14 +46,15 @@ c = Coverage(
         axes={
             "x": {"dataType": "float", "values": [1.23]},
             "y": {"values": [4.56]},
-            "t": {"dataType": "datetime", "values": [datetime.datetime.now()]}
+            "t": {"dataType": "datetime", "values": [datetime.now(tz=timezone.utc)]},
```
Comment (lukas-phaf) [inline README.md:49]: I missed this before:
This is not the intended use of datatype, it should be left out in this case. 
See the 4th and 5th bullet in section [9.6.1.1](https://opengeospatial.github.io/ogcna-auto-review/21-069.html#_7770268a-bf17-481a-ac40-b309e60548a3). In this case `dataType` should be set to `primitive` and thus left out.

The same issue is in the example.

There is currently no enum to enforce this, as that will block custom extensions...

Comment (PaulVanSchayck) [reply README.md]: Fixed

File: pyproject.toml [hunk 2]
```diff
@@ -19,8 +19,8 @@ classifiers = [
     "Topic :: Scientific/Engineering :: GIS",
     "Typing :: Typed",
 ]
-version = "0.1.0"
-dependencies = ["pydantic>=1,<2", "orjson>=3"]
+version = "0.2.0"
+dependencies = ["pydantic>=2,<3"]
```
Comment (lukas-phaf) [inline pyproject.toml:23]: Put >=2.3 here, as that is what we know it works against?

File: src/covjson_pydantic/base_models.py [hunk 3]
```diff
@@ -1,32 +1,16 @@
...
-    class Config:
-        anystr_strip_whitespace = True
-        min_anystr_length = 1
-        extra = Extra.forbid
-        validate_all = True
-        validate_assignment = True
-
-        json_loads = orjson.loads
-        json_dumps = orjson_dumps
+    model_config = ConfigDict(
```
Comment (lukas-phaf) [inline src/covjson_pydantic/base_models.py:6]: Did you review [strict](https://docs.pydantic.dev/latest/usage/strict_mode/) mode, and if that is a better option for us?

File: src/covjson_pydantic/base_models.py [hunk 4]
```diff
@@ -1,32 +1,16 @@
...
+        extra="forbid",
+        validate_default=True,
+        validate_assignment=True,
+    )
 
 
 class CovJsonBaseModel(BaseModel):
-    class Config:
-        extra = Extra.allow  # allow custom members
+    model_config = ConfigDict(extra="allow")
```
Comment (lukas-phaf) [inline src/covjson_pydantic/base_models.py:16]: Previously, this kept all the other config setting in BaseModel, only overwriting the `extra` setting. Now I am not so sure...

File: src/covjson_pydantic/base_models.py [hunk 5]
```diff
@@ -1,32 +1,16 @@
...
-        extra = Extra.forbid
-        validate_all = True
-        validate_assignment = True
-
-        json_loads = orjson.loads
-        json_dumps = orjson_dumps
+    model_config = ConfigDict(
+        str_strip_whitespace=True,
+        str_min_length=1,
+        extra="forbid",
```
Comment (lukas-phaf) [inline src/covjson_pydantic/base_models.py:9]: Why not use the Enum explicitly like before?

File: src/covjson_pydantic/domain.py [hunk 6]
```diff
@@ -23,34 +22,30 @@ class CompactAxis(BaseModel):
...
+        return self
 
 
 ValuesT = TypeVar("ValuesT")
 
 
-class ValuesAxis(GenericModel, Generic[ValuesT]):
-    dataType: Optional[str]  # noqa: N815
-    coordinates: Optional[List[str]]
+class ValuesAxis(BaseModel, Generic[ValuesT]):
```
Comment (lukas-phaf) [inline src/covjson_pydantic/domain.py:35]: Looking at this again, pydantic `GenericModel` was really intended here, as it this object a local config.

But I think we get the same config if we use based model `CovJsonBaseModel` here, and drop the local config, agreed?

File: src/covjson_pydantic/base_models.py [hunk 7]
```diff
@@ -1,32 +1,16 @@
...
+        validate_default=True,
+        validate_assignment=True,
+    )
 
 
 class CovJsonBaseModel(BaseModel):
```
Comment (lukas-phaf) [inline src/covjson_pydantic/base_models.py:15]: Originally the naming was `CovjsonBasedModel` for the base model for the "top level" CovJson object, and `BaseModel` for the rest. But I don't think that really applies anymore, so maybe `BaseModelNoExtra` and `BaseModelWithExtra` are clearer now?

File: src/covjson_pydantic/i18n.py [hunk 8]
```diff
@@ -9,4 +9,10 @@ class LanguageTag(str, Enum):
     undefined = "und"
 
 
-i18n = Dict[LanguageTag, str]
+# TODO: This was throwing warning:
+#   Expected `definition-ref` but got `LanguageTag` - serialized value may not be as expected
+# This may be a bug in Pydantic: https://github.com/pydantic/pydantic/issues/6467
+# or: https://github.com/pydantic/pydantic/issues/6422
+# So, for now, reverted to a less strict type
+# i18n = Dict[LanguageTag, str]
```
Comment (lukas-phaf) [inline src/covjson_pydantic/i18n.py:17]: Hmm, annoying. Do we want to track this with a ticket?

File: src/covjson_pydantic/ndarray.py [hunk 9]
```diff
@@ -19,39 +18,35 @@ class DataType(str, Enum):
...
                 )
 
-        return values
+        return self
 
 
 class TileSet(BaseModel):
     tileShape: List[Optional[int]]  # noqa: N815
-    urlTemplate: AnyUrl  # noqa: N815
+    urlTemplate: str  # noqa: N815
```
Comment (lukas-phaf) [inline src/covjson_pydantic/ndarray.py:49]: Is this change needed?

File: src/covjson_pydantic/base_models.py (hunk 7 above)
Comment (PaulVanSchayck) [reply src/covjson_pydantic/base_models.py]: Agreed, your second option is probably cleanest, and easiest to maintain. Implemented.

File: src/covjson_pydantic/base_models.py (hunk 4 above)
Comment (PaulVanSchayck) [reply src/covjson_pydantic/base_models.py]: No.. it was not needed (anymore). Changed.

File: src/covjson_pydantic/ndarray.py (hunk 9 above)
Comment (PaulVanSchayck) [reply src/covjson_pydantic/ndarray.py:48]: Yes. See example in `spec-tiled-ndarray.json`. 
```
"urlTemplate": "http://example.com/c/{y}-{x}.covjson"
```
When you use AnyUrl. You get the following error:
```
E       assert '{"type":"Til...D.covjson"}]}' == '{"type":"Til...}.covjson"}]}'
E         Skipping 230 identical leading characters in diff, use -v to show
E         - ple.com/b/{t}.covjson"},{"tileShape":[null,2,3],"urlTemplate":"http://example.com/c/{y}-{x}.covjson"}]}
E         ?           ^ ^                                                                       ^ ^ ^ ^
E         + ple.com/b/%7Bt%7D.covjson"},{"tileShape":[null,2,3],"urlTemplate":"http://example.com/c/%7By%7D-%7Bx%7D.covjson"}]}
E         ?           ^^^ ^^^                                                                       ^^^ ^^^ ^^^ ^^^
```
So, it's URL encoding in between. But we don't want this here, because it's technically not an URL yet. It's a template.

File: src/covjson_pydantic/domain.py [hunk 10]
```diff
@@ -62,31 +52,25 @@ class DomainType(str, Enum):
...
+
+    @model_validator(mode="after")
+    def at_least_one_axes(self):
+        if self.x is None and self.y is None and self.z is None and self.t is None and self.composite is None:
             raise ValueError("At least one axis of x,y,z,t or composite must be given.")
-        return values
+        return self
 
 
 class Domain(CovJsonBaseModel):
```
Comment (lukas-phaf) [inline src/covjson_pydantic/domain.py:69]: This used to have an extra=allow, because it used CovJsonBaseModel. This is also in the spec. At least that is how I interpret [this](https://opengeospatial.github.io/ogcna-auto-review/21-069.html#_070a1d4d-ff49-426b-8f6a-0c40fd059408)

File: src/covjson_pydantic/base_models.py (hunk 3 above)
Comment (lukas-phaf) [reply src/covjson_pydantic/base_models.py:6]: Ik neig naar strict mode aan zetten, en als dat voor sommige velden lastig is, het weer uitzetten voor dat specifiek veld. Zie de note in deze sectie:
https://docs.pydantic.dev/latest/usage/strict_mode/#strict-mode-with-configdict

File: src/covjson_pydantic/domain.py (hunk 10 above)
Comment (PaulVanSchayck) [reply src/covjson_pydantic/domain.py]: Woops.. missed that one in refactor

File: src/covjson_pydantic/i18n.py (hunk 8 above)
Comment (PaulVanSchayck) [reply src/covjson_pydantic/i18n.py]: Yes: https://github.com/KNMI/covjson-pydantic/issues/3

File: src/covjson_pydantic/base_models.py (hunk 3 above)
Comment (PaulVanSchayck) [reply src/covjson_pydantic/base_models.py:6]: Strict mode enabled, except for ValuesAxis

**Review by lukas-phaf — APPROVED** [review]
